File: src/integration/wiremock/src/jvmMain/java/community/flock/wirespec/integration/wiremock/java/WirespecWireMock.java
package community.flock.wirespec.integration.wiremock.java;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.tomakehurst.wiremock.client.MappingBuilder;
import com.github.tomakehurst.wiremock.client.ResponseDefinitionBuilder;
import com.github.tomakehurst.wiremock.client.WireMock;
import com.github.tomakehurst.wiremock.matching.UrlPattern;
import community.flock.wirespec.integration.jackson.java.WirespecSerialization;
import community.flock.wirespec.java.Wirespec;

import java.util.Map;
import java.util.regex.Pattern;

/**
 * Start building a WireMock stub for a Wirespec endpoint. Mirrors WireMock's own
 * {@code get(urlEqualTo(...))} / {@code post(urlEqualTo(...))} factories — the returned
 * builder accepts only {@link Wirespec.Response} values that belong to the same endpoint:
 *
 * <pre>
 *     import static community.flock.wirespec.integration.wiremock.java.WirespecWireMock.wirespec;
 *
 *     server.stubFor(wirespec(new GetTodos.Handler.Handlers())
 *             .willReturn(new GetTodos.Response200(todos)));
 * </pre>
 *
 * Passing a response from a different endpoint is a compile error.
 *
 * The endpoint's HTTP method and path template drive the WireMock request matcher
 * (path parameters match any non-slash segment).
 */
public final class WirespecWireMock {

    private static final Pattern PATH_PARAM_PATTERN = Pattern.compile("\\{[^/}]+}");
    private static final Wirespec.Serialization DEFAULT_SERIALIZATION = new WirespecSerialization(new ObjectMapper());

    private WirespecWireMock() {}

    public static <Req extends Wirespec.Request<?>, Res extends Wirespec.Response<?>> WirespecMappingBuilder<Res> wirespec(
            Wirespec.Server<Req, Res> endpoint) {
        return new WirespecMappingBuilder<>(endpoint, requestBuilder(endpoint));
    }

    public static final class WirespecMappingBuilder<Res extends Wirespec.Response<?>> {
        private final Wirespec.Server<?, Res> endpoint;
        private final MappingBuilder mapping;

        WirespecMappingBuilder(Wirespec.Server<?, Res> endpoint, MappingBuilder mapping) {
            this.endpoint = endpoint;
            this.mapping = mapping;
        }

        /**
         * Serialize {@code response} through a Jackson-backed {@link Wirespec.Serialization}
         * and attach it as this stub's response. Returns the underlying {@link MappingBuilder}
         * so callers can keep chaining WireMock methods (e.g. {@code .atPriority(...)},
         * {@code .inScenario(...)}).
         */
        public MappingBuilder willReturn(Res response) {
            return willReturn(response, DEFAULT_SERIALIZATION);
        }

        /**
         * Serialize {@code response} through {@code serialization} and attach it as this stub's response.
         * Pass your own {@link Wirespec.Serialization} to customize the ObjectMapper or swap in a
         * different serializer.
         */
        public MappingBuilder willReturn(Res response, Wirespec.Serialization serialization) {
            return mapping.willReturn(responseBuilder(endpoint.getServer(serialization).to(response)));
        }
    }

    private static MappingBuilder requestBuilder(Wirespec.Server<?, ?> endpoint) {
        UrlPattern urlPattern = urlPatternFor(endpoint.getPathTemplate());
        return switch (endpoint.getMethod().toUpperCase()) {
            case "GET" -> WireMock.get(urlPattern);
            case "PUT" -> WireMock.put(urlPattern);
            case "POST" -> WireMock.post(urlPattern);
            case "DELETE" -> WireMock.delete(urlPattern);
            case "PATCH" -> WireMock.patch(urlPattern);
            case "HEAD" -> WireMock.head(urlPattern);
            case "OPTIONS" -> WireMock.options(urlPattern);
            case "TRACE" -> WireMock.trace(urlPattern);
            default -> WireMock.any(urlPattern);
        };
    }

    private static ResponseDefinitionBuilder responseBuilder(Wirespec.RawResponse rawResponse) {
        ResponseDefinitionBuilder builder = WireMock.aResponse().withStatus(rawResponse.statusCode());
        for (Map.Entry<String, java.util.List<String>> entry : rawResponse.headers().entrySet()) {
            for (String value : entry.getValue()) {
                builder.withHeader(entry.getKey(), value);
            }
        }
        rawResponse.body().ifPresent(builder::withBody);
        return builder;
    }

    static UrlPattern urlPatternFor(String pathTemplate) {
        if (!PATH_PARAM_PATTERN.matcher(pathTemplate).find()) {
            return WireMock.urlPathEqualTo(pathTemplate);
        }
        String[] parts = PATH_PARAM_PATTERN.split(pathTemplate, -1);
        StringBuilder regex = new StringBuilder();
        for (int i = 0; i < parts.length; i++) {
            regex.append(Pattern.quote(parts[i]));
            if (i < parts.length - 1) {
                regex.append("[^/]+");
            }
        }
        return WireMock.urlPathMatching(regex.toString());
    }
}

File: src/integration/wiremock/src/jvmMain/kotlin/community/flock/wirespec/integration/wiremock/kotlin/WirespecWireMock.kt
package community.flock.wirespec.integration.wiremock.kotlin

import com.fasterxml.jackson.databind.ObjectMapper
import com.github.tomakehurst.wiremock.client.MappingBuilder
import com.github.tomakehurst.wiremock.client.ResponseDefinitionBuilder
import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.http.Request
import com.github.tomakehurst.wiremock.matching.UrlPattern
import community.flock.wirespec.integration.jackson.v2.kotlin.WirespecSerialization
import community.flock.wirespec.kotlin.Wirespec
import java.net.URI
import java.net.URLDecoder

/**
 *
 * ```
 * server.stubFor(wirespec(GetTodos.Handler).willReturn(GetTodos.Response200(todos)))
 * ```
 */
fun <Req : Wirespec.Request<*>, Res : Wirespec.Response<*>> wirespec(
    endpoint: Wirespec.Server<Req, Res>,
): WirespecMappingBuilder<Res> = WirespecMappingBuilder(endpoint, requestBuilder(endpoint.method, endpoint.pathTemplate))

class WirespecMappingBuilder<Res : Wirespec.Response<*>> internal constructor(
    private val endpoint: Wirespec.Server<*, Res>,
    private val mapping: MappingBuilder,
) {
    /**
     */
    fun willReturn(
        response: Res,
        serialization: Wirespec.Serialization = defaultSerialization,
    ): MappingBuilder = mapping.willReturn(responseBuilder(endpoint.server(serialization).to(response)))
}

private val defaultSerialization: Wirespec.Serialization by lazy { WirespecSerialization(ObjectMapper()) }

fun requestBuilder(method: String, pathTemplate: String): MappingBuilder {
    val urlPattern = urlPatternFor(pathTemplate)
    return when (method.uppercase()) {
        "GET" -> WireMock.get(urlPattern)
        "PUT" -> WireMock.put(urlPattern)
        "POST" -> WireMock.post(urlPattern)
        "DELETE" -> WireMock.delete(urlPattern)
        "PATCH" -> WireMock.patch(urlPattern)
        "HEAD" -> WireMock.head(urlPattern)
        "OPTIONS" -> WireMock.options(urlPattern)
        "TRACE" -> WireMock.trace(urlPattern)
        else -> WireMock.any(urlPattern)
    }
}

fun responseBuilder(rawResponse: Wirespec.RawResponse): ResponseDefinitionBuilder {
    val builder = WireMock.aResponse().withStatus(rawResponse.statusCode)
    rawResponse.headers.forEach { (name, values) ->
        values.forEach { value -> builder.withHeader(name, value) }
    }
    rawResponse.body?.let(builder::withBody)
    return builder
}

fun Request.toRawRequest(): Wirespec.RawRequest {
    val uri = URI.create(absoluteUrl)
    val segments = uri.rawPath.split("/").filter(String::isNotEmpty).map(::decode)
    val queries = (uri.rawQuery ?: "").split("&").filter(String::isNotEmpty)
        .map { it.split("=", limit = 2) }
        .groupBy({ decode(it[0]) }, { decode(it.getOrElse(1) { "" }) })
    return Wirespec.RawRequest(
        method = method.value(),
        path = segments,
        queries = queries,
        headers = headers.all().associate { it.key() to it.values().toList() },
        body = body?.takeIf(ByteArray::isNotEmpty),
    )
}

private fun decode(value: String): String = URLDecoder.decode(value, Charsets.UTF_8)

private val PATH_PARAM_REGEX = Regex("""\{[^/}]+\}""")

internal fun urlPatternFor(pathTemplate: String): UrlPattern = if (PATH_PARAM_REGEX.containsMatchIn(pathTemplate)) {
    val regex = pathTemplate.split(PATH_PARAM_REGEX).joinToString("[^/]+") { Regex.escape(it) }
    WireMock.urlPathMatching(regex)
} else {
    WireMock.urlPathEqualTo(pathTemplate)
}
